src/main/java/org/broadinstitute/hellbender/tools/exome/DiscoverCopyNumberTriStateSegments.java

```diff
@@ -0,0 +1,484 @@
+        CopyNumberTriState currentSegmentState = bestPath.get(0);
+        Target lastTargetInPreviousSegment = null;
+        currentSegmentTargets.add(targets.get(0));
+        currentSegmentData.add(coverage.get(0));
+        currentSegmentInputValues.add(inputValues.get(0));
+        for (int i = 1; i < targets.size(); i++) {
+            final CopyNumberTriState thisState = bestPath.get(i);
+            final Double thisDatum = coverage.get(i);
+            final Target target = targets.get(i);
+            if (currentSegmentState == thisState
+                    && target.getContig().equals(currentSegmentTargets.get(0).getContig())) {
+                currentSegmentTargets.add(target);
+                currentSegmentData.add(thisDatum);
+                currentSegmentInputValues.add(inputValues.get(i));
+            } else {
+                result.add(composeSegment(currentSegmentTargets, currentSegmentState,
+                                          lastTargetInPreviousSegment, target, fbResult, currentSegmentInputValues));
+                lastTargetInPreviousSegment = currentSegmentTargets.get(currentSegmentTargets.size() - 1);
+                currentSegmentTargets.clear();
+                currentSegmentTargets.add(target);
+                currentSegmentData.clear();
+                currentSegmentData.add(thisDatum);
+                currentSegmentInputValues.clear();
+                currentSegmentInputValues.add(inputValues.get(i));
+                currentSegmentState = thisState;
+            }
+        }
+        result.add(composeSegment(currentSegmentTargets, currentSegmentState,
+                                  lastTargetInPreviousSegment, null, fbResult, currentSegmentInputValues));
+        return result;
+    }
+
+    /**
+     * Composes the segment calculating all the corresponding quality scores,
+     *  coverage mean and stddev.
+     *
+     * @param targets the targets enclosed in the segment.
+     * @param call the segment called state.
+     * @param precedingTarget the target preceding the segment's start target.
+     * @param followingTarget the target following the segment's last target.
+     * @param fbResult the result of running the Forward-Backward algorithm.
+     * @return never {@code null}.
+     */
+    private CopyNumberTriStateSegment composeSegment(final List<Target> targets,
+                                                     final CopyNumberTriState call,
+                                                     final Target precedingTarget, final Target followingTarget,
+                                                     final ForwardBackwardAlgorithm.Result<Double, Target, CopyNumberTriState> fbResult,
+                                                     final List<Double> inputValues) {
+        final int length = targets.size();
+        final Target firstTarget = targets.get(0);
+        final double mean = inputValues.stream().mapToDouble(d -> d).sum() / length;
```

Comment:
`.average().getAsDouble()` has an annoying unboxing from the `OptionalDouble` but IMO it's still preferable because it conveys the point of the calculation more directly.